doc/SAI_Proposal_Generic_Extensions.md

```diff
@@ -0,0 +1,423 @@
+
+Title       | SAI Generic Extensions Proposal
+------------|--------------------------------
+Authors     | Intel.
+Status      | Draft
+Type        | Standards track
+Created     | 08/04/2022
+SAI-Version | 1.11
+---------------------------------------------
+
+## Table of Contents
+- [Scope](#scope)
+- [Overview](#overview)
+  - [sai_json_t](#saijsont)
+  - [SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE](#saiobjecttypegenericprogrammable)
+- [SAI pipeline](#sai-pipeline)
+- [An Example - L4LB](#an-example---l4lb)
+  - [JSON data](#json-data)
+  - [Counters](#counters)
+  - [CRM](#crm)
+- [API specification](#api-specification)
+  - [New header file saigenericextensions.h](#new-header-file-saigenericextensionsh)
+- [API usage](#api-usage)
+  - [Create](#create)
+  - [Update](#update)
+  - [Delete](#delete)
+  - [Object Query](#object-query)
+  - [CRM API](#crm-usage)
+- [Usage Examples for sai_json_t](#usage-examples-for-saijsont)
+  - [Refer to an OID - Port](#refer-to-an-oid---port)
+  - [Refer to an OID - Generic Programmable Block](#refer-to-an-oid---generic-programmable-block)
+  - [Refer to a key based object - Route](#refer-to-a-key-based-object---route)
+
+# Scope #
+This document defines the technical specification for the API to support SAI Extensions. Extensions are an abstraction over niche/device-specific features. They facilitate the configuration of newly supported features of ASICs via the familiar SAI API model. The exact feature enabled via the extensions itself is outside the scope of this document but an example is provided below to better explain this feature. These extensions are not a replacement to the existing SAI objects but instead provide a complementary path to achieve the goals set out below.
+
+# Overview #
+SAI Generic extensions introduce a new SAI object SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE of object_type_oid. This new object uses a single attribute of type sai_json_t defined in this proposal. This object and its attribute provide an abstraction of target-specific configurable features otherwise not described in SAI. This new object enables the 
+- Support ability to add user features
+- Support device-specific capabilities
+- Rapid application prototyping
+
+The purpose of this document is to introduce these new entities and how they can be leveraged to extend data plane functionality.
+1. sai_json_t - New SAI datatype 
+2. SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE - New SAI object type
+
+## sai_json_t ##
+sai_json_t defines a generic data type using JSON as the data format with a predefined grammar allowed to be used within. This grammar allows validators like SAI metadata to verify data and corresponding types accurately. The user of this data type copies a JSON string into this type. The exact structure or content of the API is opaque and is considered an agreed contract between the application and SAI implementation.
+
+```c
+/**
+ * @brief
+\verbatim
+"attributes": [
+    {
+        <attribute_name>: {
```

Comment:
Please specify how the SAI application knows the h/w table (vip_table in this example) and the attributes of the h/w table (virtual_ip and nh_list). 

We should avoid making this vendor specific. Doesn't this make the SAI application locked-in to a particular SAI vendor?